{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c72f4cbf",
   "metadata": {},
   "source": [
    "# TreeAlign with allele specfic information\n",
    "## Introduction\n",
    "TreeAlign is a model for scRNA and scDNA integration. TreeAlign can take in either total copy number information or allele specific copy number information or both to assign cells from scRNA to a phylogenetic tree constructed with scDNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c3d1f05",
   "metadata": {},
   "source": [
    "## Loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a2490419",
   "metadata": {},
   "outputs": [],
   "source": [
    "from treealign import CloneAlignClone\n",
    "from treealign import CloneAlignTree\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Phylo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cf2508c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load total copy number input\n",
    "\n",
    "# scRNA read count matrix where each row represents a gene, \n",
    "# each column represents a cell\n",
    "expr = pd.read_csv(\"../data/example_expr.csv\", index_col=0)\n",
    "\n",
    "# scDNA copy number matrix where each row represents a gene,\n",
    "# each column represents a cell\n",
    "cnv = pd.read_csv(\"../data/example_gene_cnv.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "68319b0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load allele specific input\n",
    "\n",
    "# b allele frequency matrix\n",
    "# each row represents a snp\n",
    "# each column represents a cell\n",
    "# The number in the matrix is the b allele frequency at the given snp and cell\n",
    "hscn = pd.read_csv(\"../data/example_snp_baf.csv\", index_col=0)\n",
    "\n",
    "# reference allele count matrix from scRNA\n",
    "# each row represents a snp\n",
    "# each column represents a cell\n",
    "snv_allele = pd.read_csv(\"../data/example_snp_allele.csv\", index_col=0)\n",
    "\n",
    "# total count matrix at SNPs from scRNA\n",
    "# each row represents a snp\n",
    "# each column represents a cell\n",
    "snv_total = pd.read_csv(\"../data/example_snp_total.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8883c8b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load phylogenetic tree used for clone assignment\n",
    "\n",
    "# phylogenetic tree constructed with scDNA data in newick format\n",
    "tree = Phylo.read(\"../data/example_hdbscan.newick\", \"newick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01a19d88",
   "metadata": {},
   "source": [
    "## Running TreeAlign with tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "364a76a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/juno/work/shah/users/weinera2/projects/TreeAlign/venv/lib/python3.7/site-packages/treealign/clonealign.py:64: FutureWarning: Index.__and__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__and__.  Use index.intersection(other) instead\n",
      "  intersect_index = self.hscn_df.index & self.snv_allele_df.index & self.snv_df.index\n",
      "/juno/work/shah/users/weinera2/projects/TreeAlign/venv/lib/python3.7/site-packages/treealign/clonealign.py:66: FutureWarning: Index.__and__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__and__.  Use index.intersection(other) instead\n",
      "  intersect_cells = self.snv_allele_df.columns & self.snv_df.columns\n",
      "/juno/work/shah/users/weinera2/projects/TreeAlign/venv/lib/python3.7/site-packages/treealign/clonealign.py:77: FutureWarning: Index.__and__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__and__.  Use index.intersection(other) instead\n",
      "  intersect_cells = self.expr_df.columns & self.snv_allele_df.columns & self.snv_df.columns\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_0, one of the child clade is node_1 with 341 terminals. \n",
      "At node_0, one of the child clade is node_341 with 721 terminals. \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/juno/work/shah/users/weinera2/projects/TreeAlign/venv/lib/python3.7/site-packages/treealign/clonealign.py:146: FutureWarning: Index.__and__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__and__.  Use index.intersection(other) instead\n",
      "  intersect_index = clone_hscn_df.index & snv.index & snv_allele.index\n",
      "/juno/work/shah/users/weinera2/projects/TreeAlign/venv/lib/python3.7/site-packages/pyro/util.py:303: UserWarning: Found vars in model but not guide: {'allele', 'gene_type'}\n",
      "  warnings.warn(f\"Found vars in model but not guide: {bad_sites}\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1371 genes in matrices. \n",
      "Start run clonealign for clade: node_0\n",
      "cnv gene count: 1371\n",
      "expr cell count: 1000\n",
      "hscn snp count: 1355\n",
      "snv allele matrix cell count: 1000\n",
      "seed = 55, initial_loss = 6514374.023016376\n",
      "Start Inference.\n",
      "ELBO converged at iteration 75\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_0 with correct frequency 0.995\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_1, one of the child clade is node_2 with 77 terminals. \n",
      "At node_1, one of the child clade is node_78 with 264 terminals. \n",
      "There are 1579 genes in matrices. \n",
      "Start run clonealign for clade: node_1\n",
      "cnv gene count: 1579\n",
      "expr cell count: 266\n",
      "hscn snp count: 1291\n",
      "snv allele matrix cell count: 266\n",
      "seed = 55, initial_loss = 2091303.4187890508\n",
      "Start Inference.\n",
      "ELBO converged at iteration 246\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_1 with correct frequency 1.0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_2, there are less than 20 cells in the expr matrix.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_78, one of the child clade is node_79 with 32 terminals. \n",
      "At node_78, one of the child clade is node_110 with 232 terminals. \n",
      "There are 355 genes in matrices. \n",
      "Start run clonealign for clade: node_78\n",
      "cnv gene count: 355\n",
      "expr cell count: 263\n",
      "hscn snp count: 208\n",
      "snv allele matrix cell count: 263\n",
      "seed = 98, initial_loss = 511241.8064454052\n",
      "Start Inference.\n",
      "ELBO converged at iteration 79\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_78 with correct frequency 0.9923954372623575\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_79, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_80, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_81, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_82, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_83, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_84, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_85, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_86, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_87, the level limit exceeds.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_110, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_111, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_112, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_113, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_114, one of the child clade is node_115 with 32 terminals. \n",
      "At node_114, one of the child clade is node_146 with 196 terminals. \n",
      "There are 192 genes in matrices. \n",
      "Start run clonealign for clade: node_114\n",
      "cnv gene count: 192\n",
      "expr cell count: 180\n",
      "hscn snp count: 73\n",
      "snv allele matrix cell count: 180\n",
      "seed = 55, initial_loss = 185605.3237396796\n",
      "Start Inference.\n",
      "ELBO converged at iteration 39\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_114 with correct frequency 0.9166666666666666\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_115, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_116, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_117, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_118, the level limit exceeds.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_146, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_147, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_148, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_149, the level limit exceeds.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_341, one of the child clade is node_342 with 50 terminals. \n",
      "At node_341, one of the child clade is node_391 with 671 terminals. \n",
      "There are 959 genes in matrices. \n",
      "Start run clonealign for clade: node_341\n",
      "cnv gene count: 959\n",
      "expr cell count: 729\n",
      "hscn snp count: 508\n",
      "snv allele matrix cell count: 729\n",
      "seed = 98, initial_loss = 2468684.74629217\n",
      "Start Inference.\n",
      "ELBO converged at iteration 66\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_341 with correct frequency 0.9711934156378601\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_342, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_343, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_344, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_345, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_346, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_347, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_348, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_349, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_350, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_351, the level limit exceeds.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_391, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_392, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_393, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_394, one of the child clade is node_395 with 63 terminals. \n",
      "At node_394, one of the child clade is node_457 with 605 terminals. \n",
      "There are 640 genes in matrices. \n",
      "Start run clonealign for clade: node_394\n",
      "cnv gene count: 640\n",
      "expr cell count: 584\n",
      "hscn snp count: 608\n",
      "snv allele matrix cell count: 584\n",
      "seed = 50, initial_loss = 1244902.0286313116\n",
      "Start Inference.\n",
      "ELBO converged at iteration 283\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_394 with correct frequency 1.0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_395, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_396, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_397, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_398, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_399, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_400, the level limit exceeds.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_457, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_458, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_459, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_460, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_461, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_462, the level limit exceeds.\n"
     ]
    }
   ],
   "source": [
    "# construct CloneAlignTree object for data preprocessing\n",
    "# run TreeAlign with both total copy number & allele specific datasets\n",
    "\n",
    "# `repeat` is set to 1 here for demonstration purposes. it would be better to set `repeat` larger than 5. \n",
    "obj = CloneAlignTree(tree=tree, expr=expr, cnv=cnv, hscn=hscn, snv_allele=snv_allele, snv=snv_total, repeat=1)\n",
    "\n",
    "# it is possible to run TreeAlign with total copy number data only\n",
    "# obj = CloneAlignTree(tree=tree, expr=expr, cnv=cnv, repeat=1)\n",
    "\n",
    "# it is also possible to run TreeAlign with allele specific data only\n",
    "# obj = CloneAlignTree(tree=tree, hscn=hscn, snv_allele=snv_allele, snv=snv_total, repeat=1)\n",
    "\n",
    "# running TreeAlign to assign cells to phylogenetic subclades\n",
    "obj.assign_cells_to_tree()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "71027764",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class CloneAlignTree in module treealign.clonealign_tree:\n",
      "\n",
      "class CloneAlignTree(treealign.clonealign.CloneAlign)\n",
      " |  CloneAlignTree(tree, expr=None, cnv=None, hscn=None, snv_allele=None, snv=None, normalize_cnv=True, cnv_cutoff=10, infer_s_score=True, infer_b_allele=True, repeat=10, min_clone_assign_prob=0.8, min_clone_assign_freq=0.7, min_consensus_gene_freq=0.6, min_consensus_snv_freq=0.6, max_temp=1.0, min_temp=0.5, anneal_rate=0.01, learning_rate=0.1, max_iter=400, rel_tol=5e-05, record_input_output=False, min_cell_count_expr=20, min_cell_count_cnv=20, min_gene_diff=100, min_snp_diff=100, level_cutoff=10, min_proceed_freq=0.7, min_record_freq=0.7)\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      CloneAlignTree\n",
      " |      treealign.clonealign.CloneAlign\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, tree, expr=None, cnv=None, hscn=None, snv_allele=None, snv=None, normalize_cnv=True, cnv_cutoff=10, infer_s_score=True, infer_b_allele=True, repeat=10, min_clone_assign_prob=0.8, min_clone_assign_freq=0.7, min_consensus_gene_freq=0.6, min_consensus_snv_freq=0.6, max_temp=1.0, min_temp=0.5, anneal_rate=0.01, learning_rate=0.1, max_iter=400, rel_tol=5e-05, record_input_output=False, min_cell_count_expr=20, min_cell_count_cnv=20, min_gene_diff=100, min_snp_diff=100, level_cutoff=10, min_proceed_freq=0.7, min_record_freq=0.7)\n",
      " |      initialize CloneAlignTree object\n",
      " |      :param expr: expr read count matrix. row is gene, column is cell. (pandas.DataFrame)\n",
      " |      :param cnv: cnv matrix. row is gene, column is cell. (pandas.DataFrame)\n",
      " |      :param tree: phylogenetic tree of cells (Bio.Phylo.BaseTree.Tree)\n",
      " |      :param normalize_cnv: whether to normalized cnv matrix by min or not. (bool)\n",
      " |      :param cnv_cutoff: set cnv higher than cnv_cutoff to cnv_cutoff. (int)\n",
      " |      :param model_select: \"gene\" for the extended clonealign model or \"default\" for the original clonelign model (str)\n",
      " |      :param repeat: num of times to run clonealign to generate consensus results. (int)\n",
      " |      :param min_cell_count_expr: min cells in scRNA to keep assigning cells to subtrees\n",
      " |      :param min_cell_count_cnv: min cells in the current subtree to proceed to the next level\n",
      " |      :param min_gene_diff: min number of genes that are different between subtrees to keep assigning\n",
      " |      :param level_cutoff: stop clonealign when get to subtrees of certain level\n",
      " |      :param min_proceed_freq: proceed clonealign to the next level if a certain frequency of cells have consistent assignments between runs\n",
      " |      :param min_clone_assign_prob: assign cells to a clone if clone assignment prob reaches min_clone_assign_prob (float)\n",
      " |      :param min_clone_assign_freq: assign cells to a clone if a min proportion of runs generate the same results (float)\n",
      " |      :param max_temp: starting temperature in Gumbel-Softmax reparameterization. (float)\n",
      " |      :param min_temp: min temperature in Gumbel-Softmax reparameterization. (float)\n",
      " |      :param anneal_rate: annealing rate in Gumbel-Softmax reparameterization. (float)\n",
      " |      :param learning_rate: learning rate of Adam optimizer. (float)\n",
      " |      :param max_iter: max number of iterations of elbo optimization during inference. (int)\n",
      " |      :param rel_tol: when the relative change in elbo drops to rel_tol, stop inference. (float)\n",
      " |  \n",
      " |  add_tree_node_name(self, node)\n",
      " |  \n",
      " |  assign_cells_to_clade(self, current_clade, expr_cells, level)\n",
      " |      assign cells to a clade in Phylo tree\n",
      " |      :param current_clade: (Bio.Phylo.BaseTree.Clade)\n",
      " |      :param expr_cells: cells from scRNA (list[str])\n",
      " |      :param level: current level of the clade\n",
      " |      :return: None\n",
      " |  \n",
      " |  assign_cells_to_tree(self)\n",
      " |      assign cells to Phylo tree\n",
      " |      :return: clone_assign_df (pandas.DataFrame) and gene_type_score_df (pandas.DataFrame)\n",
      " |  \n",
      " |  record_clone_assign_to_dict(self, expr_cells, clone_assign, clean_clades)\n",
      " |      record clone assignment results to self.clone_assign_dict\n",
      " |      :param expr_cells: cells in expr matrix (list[str])\n",
      " |      :param clone_assign: clone assignments (pandas.Series)\n",
      " |      :param clean_clades: clean clades in the current run (list[Clade])\n",
      " |      :return: None\n",
      " |  \n",
      " |  record_param_to_dict(self, param_dict, indices, params)\n",
      " |      Update params in params\n",
      " |      :param indices: gene names (list[str])\n",
      " |      :param gene_type_score: mean gene_type_score across runs (pandas.Series)\n",
      " |      :return:\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from treealign.clonealign.CloneAlign:\n",
      " |  \n",
      " |  average_param_dict(self, param_dict)\n",
      " |      # inplace method\n",
      " |  \n",
      " |  clonealign_pyro_model(self, cnv, expr, hscn, snv_allele, snv, infer_s_score=True, infer_b_allele=True, temperature=0.5)\n",
      " |      original clonealign model\n",
      " |      :param cnv: torch.tensor\n",
      " |      :param expr: torch.tensor\n",
      " |      :param hscn:  torch.tensor\n",
      " |      :param snv_allele: torch.tensor\n",
      " |      :param snv: torch.tensor\n",
      " |      :param gene_model: bool\n",
      " |      :param infer_b_allele: bool\n",
      " |      :param temperature: float\n",
      " |      :return: None\n",
      " |  \n",
      " |  construct_allele_specific_input(self, terminals, expr_cells)\n",
      " |  \n",
      " |  construct_total_copy_number_input(self, terminals, expr_cells)\n",
      " |  \n",
      " |  convert_df_to_torch(self, df)\n",
      " |  \n",
      " |  generate_output(self)\n",
      " |      generate clone_assign_df and gene_type_score_df\n",
      " |      :return: clonealign output (pandas.DataFrame)\n",
      " |  \n",
      " |  make_columns_consistent(self, *args)\n",
      " |      # inplace method\n",
      " |  \n",
      " |  max_param_dict(self, param_dict)\n",
      " |      # inplace method\n",
      " |  \n",
      " |  process_input_matrices(self)\n",
      " |      clean up clonealign input\n",
      " |      :return: None\n",
      " |  \n",
      " |  run_clonealign_pyro(self, cnv, expr, hscn, snv_allele, snv, current_repeat)\n",
      " |      clonealign inference\n",
      " |      :param cnv: clone-specific consensus copy number matrix. row is clone. column is gene. (torch.tensor)\n",
      " |      :param expr: gene expression count matrix. row is cell. column is gene. (torch.tensor)\n",
      " |      :return: clone assignment, gene_type_score\n",
      " |  \n",
      " |  run_clonealign_pyro_repeat(self, cnv_df, expr_df, hscn_df, snv_allele_df, snv_df)\n",
      " |      call run_clonealign_pyro() multiple times to generate consensus results\n",
      " |      :param cnv_df: clone-specific consensus copy number matrix. row is gene. column is clone. (pandas.DataFrame)\n",
      " |      :param expr_df: gene expression count matrix. row is gene. column is cell. (pandas.DataFrame)\n",
      " |      :return: frequency of unassigned cells, clone assignment, gene_type_score\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from treealign.clonealign.CloneAlign:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# to view more details about parameters you can customize when you run TreeAlign\n",
    "help(CloneAlignTree)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8b9c861",
   "metadata": {},
   "source": [
    "## Getting results\n",
    "The output of TreeAlign includes: 1. a table indicating the subclades to which the cells in scRNA data are assigned. 2. for each gene, a score ranging between 0 and 1 reflecting dosage effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "313fc2d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "clone_assign_df, gene_type_score_df, allele_assign_prob_df = obj.generate_output()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2193005c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>clone_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CTAGACAT...</td>\n",
       "      <td>node_462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CTCCATGA...</td>\n",
       "      <td>node_462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_TTGGGTAG...</td>\n",
       "      <td>node_351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CCCTCAAC...</td>\n",
       "      <td>node_351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_TTGTTTGA...</td>\n",
       "      <td>node_149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_AGCTACAT...</td>\n",
       "      <td>node_351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_LEFT_ADNEXA_TCCTGCAGT...</td>\n",
       "      <td>node_462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_GTTACCCC...</td>\n",
       "      <td>node_462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_LEFT_ADNEXA_ATTTCACTC...</td>\n",
       "      <td>node_462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>994</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_ATTGTTCT...</td>\n",
       "      <td>node_400</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>995 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               cell_id  clone_id\n",
       "0    SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CTAGACAT...  node_462\n",
       "1    SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CTCCATGA...  node_462\n",
       "2    SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_TTGGGTAG...  node_351\n",
       "3    SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CCCTCAAC...  node_351\n",
       "4    SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_TTGTTTGA...  node_149\n",
       "..                                                 ...       ...\n",
       "990  SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_AGCTACAT...  node_351\n",
       "991  SPECTRUM-OV-022_S1_CD45N_LEFT_ADNEXA_TCCTGCAGT...  node_462\n",
       "992  SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_GTTACCCC...  node_462\n",
       "993  SPECTRUM-OV-022_S1_CD45N_LEFT_ADNEXA_ATTTCACTC...  node_462\n",
       "994  SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_ATTGTTCT...  node_400\n",
       "\n",
       "[995 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subclade assignment for each cell in scRNA data\n",
    "clone_assign_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b09798f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TINAGL1</td>\n",
       "      <td>3.315783e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PEF1</td>\n",
       "      <td>9.254875e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COL16A1</td>\n",
       "      <td>6.666608e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPOCD1</td>\n",
       "      <td>2.565527e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL136115.1</td>\n",
       "      <td>1.004067e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3442</th>\n",
       "      <td>LINC00992</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3443</th>\n",
       "      <td>ZNF474</td>\n",
       "      <td>8.513761e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3444</th>\n",
       "      <td>KIAA1024L</td>\n",
       "      <td>7.467213e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3445</th>\n",
       "      <td>SOWAHA</td>\n",
       "      <td>1.491436e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3446</th>\n",
       "      <td>TIFAB</td>\n",
       "      <td>6.359953e-17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3447 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene  gene_type_score\n",
       "0        TINAGL1     3.315783e-01\n",
       "1           PEF1     9.254875e-01\n",
       "2        COL16A1     6.666608e-01\n",
       "3         SPOCD1     2.565527e-02\n",
       "4     AL136115.1     1.004067e-01\n",
       "...          ...              ...\n",
       "3442   LINC00992     1.000000e+00\n",
       "3443      ZNF474     8.513761e-17\n",
       "3444   KIAA1024L     7.467213e-17\n",
       "3445      SOWAHA     1.491436e-11\n",
       "3446       TIFAB     6.359953e-17\n",
       "\n",
       "[3447 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the probability of having dosage effects for each gene\n",
    "gene_type_score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eebe1ac",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
