{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TreeAlign with allele specfic information\n",
    "## Introduction\n",
    "TreeAlign is a model for scRNA and scDNA integration. TreeAlign can take in either total copy number information or allele specific copy number information or both to assign cells from scRNA to a phylogenetic tree constructed with scDNA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from treealign import CloneAlignClone\n",
    "from treealign import CloneAlignTree\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Phylo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load total copy number input\n",
    "\n",
    "# scRNA read count matrix where each row represents a gene, \n",
    "# each column represents a cell\n",
    "expr = pd.read_csv(\"../data/example_expr.csv\", index_col=0)\n",
    "\n",
    "# scDNA copy number matrix where each row represents a gene,\n",
    "# each column represents a cell\n",
    "cnv = pd.read_csv(\"../data/example_gene_cnv.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load allele specific input\n",
    "\n",
    "# b allele frequency matrix\n",
    "# each row represents a snp\n",
    "# each column represents a cell\n",
    "# The number in the matrix is the b allele frequency at the given snp and cell\n",
    "hscn = pd.read_csv(\"../data/example_snp_baf.csv\", index_col=0)\n",
    "\n",
    "# reference allele count matrix from scRNA\n",
    "# each row represents a snp\n",
    "# each column represents a cell\n",
    "snv_allele = pd.read_csv(\"../data/example_snp_allele.csv\", index_col=0)\n",
    "\n",
    "# total count matrix at SNPs from scRNA\n",
    "# each row represents a snp\n",
    "# each column represents a cell\n",
    "snv_total = pd.read_csv(\"../data/example_snp_total.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load phylogenetic tree used for clone assignment\n",
    "\n",
    "# phylogenetic tree constructed with scDNA data in newick format\n",
    "tree = Phylo.read(\"../data/example_hdbscan.newick\", \"newick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running TreeAlign with tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_0, one of the child clade is node_1 with 341 terminals. \n",
      "At node_0, one of the child clade is node_341 with 721 terminals. \n",
      "There are 0 genes in matrices. \n",
      "Start run clonealign for clade: node_0\n",
      "hscn snp count: 1355\n",
      "snv allele matrix cell count: 1000\n",
      "Start Inference.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/shih/miniconda3/envs/pyro/lib/python3.8/site-packages/torch/autograd/__init__.py:130: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /opt/conda/conda-bld/pytorch_1607369981906/work/c10/cuda/CUDAFunctions.cpp:100.)\n",
      "  Variable._execution_engine.run_backward(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ELBO converged at iteration 312\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_0 with correct frequency 1.0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_1, one of the child clade is node_2 with 77 terminals. \n",
      "At node_1, one of the child clade is node_78 with 264 terminals. \n",
      "There are 0 genes in matrices. \n",
      "Start run clonealign for clade: node_1\n",
      "hscn snp count: 1291\n",
      "snv allele matrix cell count: 238\n",
      "Start Inference.\n",
      "ELBO converged at iteration 240\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_1 with correct frequency 1.0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_2, one of the child clade is node_3 with 26 terminals. \n",
      "At node_2, one of the child clade is node_28 with 51 terminals. \n",
      "There are 0 genes in matrices. \n",
      "Start run clonealign for clade: node_2\n",
      "hscn snp count: 1242\n",
      "snv allele matrix cell count: 37\n",
      "Start Inference.\n",
      "ELBO converged at iteration 230\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_2 with correct frequency 1.0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_3, there are less than 20 cells in the expr matrix.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_28, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_29, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_30, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_31, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_32, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_33, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_34, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_35, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_36, the level limit exceeds.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_78, one of the child clade is node_79 with 32 terminals. \n",
      "At node_78, one of the child clade is node_110 with 232 terminals. \n",
      "There are 0 genes in matrices. \n",
      "Start run clonealign for clade: node_78\n",
      "hscn snp count: 208\n",
      "snv allele matrix cell count: 201\n",
      "Start Inference.\n",
      "ELBO converged at iteration 391\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_78 with correct frequency 0.9751243781094527\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_79, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_80, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_81, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_82, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_83, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_84, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_85, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_86, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_87, the level limit exceeds.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_110, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_111, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_112, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_113, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_114, one of the child clade is node_115 with 32 terminals. \n",
      "At node_114, one of the child clade is node_146 with 196 terminals. \n",
      "There are 0 genes in matrices. \n",
      "cnv gene count less than self.min_gene_diff: 0\n",
      "snp count less than self.min_snp_diff: 73\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_341, one of the child clade is node_342 with 50 terminals. \n",
      "At node_341, one of the child clade is node_391 with 671 terminals. \n",
      "There are 0 genes in matrices. \n",
      "Start run clonealign for clade: node_341\n",
      "hscn snp count: 508\n",
      "snv allele matrix cell count: 762\n",
      "Start Inference.\n",
      "ELBO converged at iteration 212\n",
      "Clonealign finished!\n",
      "CloneAlign Tree finishes at clade: node_341 with correct frequency 0.9921259842519685\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_342, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_343, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_344, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_345, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_346, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_347, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_348, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_349, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_350, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_351, the level limit exceeds.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_391, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_392, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_393, there is only one clean child clade existing.\n",
      "\n",
      "\n",
      "\n",
      "Start processing \n",
      "At node_394, one of the child clade is node_395 with 63 terminals. \n",
      "At node_394, one of the child clade is node_457 with 605 terminals. \n",
      "There are 0 genes in matrices. \n",
      "Start run clonealign for clade: node_394\n",
      "hscn snp count: 608\n",
      "snv allele matrix cell count: 578\n",
      "Start Inference.\n",
      "ELBO converged at iteration 7\n",
      "Clonealign finished!\n",
      "CloneAlign Tree stops at clade: node_394 with correct frequency 0.6072664359861591\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# construct CloneAlignTree object for data preprocessing\n",
    "# run TreeAlign with both total copy number & allele specific datasets\n",
    "\n",
    "# `repeat` is set to 1 here for demonstration purposes. it would be better to set `repeat` larger than 5. \n",
    "obj = CloneAlignTree(tree=tree, expr=expr, cnv=cnv, hscn=hscn, snv_allele=snv_allele, snv=snv_total, repeat=1)\n",
    "\n",
    "# it is possible to run TreeAlign with total copy number data only\n",
    "# obj = CloneAlignTree(tree=tree, expr=expr, cnv=cnv, repeat=1)\n",
    "\n",
    "# it is also possible to run TreeAlign with allele specific data only\n",
    "# obj = CloneAlignTree(tree=tree, hscn=hscn, snv_allele=snv_allele, snv=snv_total, repeat=1)\n",
    "\n",
    "# running TreeAlign to assign cells to phylogenetic subclades\n",
    "obj.assign_cells_to_tree()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# to view more details about parameters you can customize when you run TreeAlign\n",
    "help(CloneAlignTree)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting results\n",
    "The output of TreeAlign includes: 1. a table indicating the subclades to which the cells in scRNA data are assigned. 2. for each gene, a score ranging between 0 and 1 reflecting dosage effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clone_assign_df, gene_type_score_df, allele_assign_prob_df = obj.generate_output()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subclade assignment for each cell in scRNA data\n",
    "clone_assign_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the probability of having dosage effects for each gene\n",
    "gene_type_score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
